File: ios/JessicaCore/Sources/JessicaCore/JessicaCore.swift
import Foundation

/// Idiomatic Swift façade over the Rust mobile core.
///
/// UI code talks only to this class; ``JessicaFFI`` stays internal.
///
/// M0 fallback: when ``JessicaFFI/makeHandle()`` returns nil (no Rust
/// library available), the façade runs a tiny pure-Swift matcher so
/// the UI can exercise the API contract end-to-end on a dev mac.
public final class JessicaCore: @unchecked Sendable {

    private let lock = NSLock()
    private var handle: OpaquePointer?
    private var fallback: PureSwiftIntents?

    public init() {
        if let h = JessicaFFI.makeHandle() {
            self.handle = h
        } else {
            self.fallback = PureSwiftIntents()
        }
    }

    deinit {
        JessicaFFI.free(handle)
    }

    /// Load the YAML intent catalogue. Returns `true` on success.
    @discardableResult
    public func loadIntents(_ yaml: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        if let fallback {
            return fallback.load(yaml)
        }
        return JessicaFFI.loadIntents(handle, yaml: yaml) == JESSICA_OK
    }

    /// Match a transcript against the loaded catalogue. Returns nil when no intent fires.
    public func matchIntent(transcript: String, language: String) -> IntentMatch? {
        lock.lock(); defer { lock.unlock() }
        if let fallback {
            return fallback.match(transcript: transcript, language: language)
        }
        guard let json = JessicaFFI.matchIntent(handle, transcript: transcript, language: language),
              let data = json.data(using: .utf8) else {
            return nil
        }
        return try? JSONDecoder().decode(IntentMatch.self, from: data)
    }

    /// How many intents are currently loaded.
    public func intentCount() -> Int64 {
        lock.lock(); defer { lock.unlock() }
        if let fallback {
            return Int64(fallback.count)
        }
        return JessicaFFI.intentCount(handle)
    }
}

File: android/core/src/main/kotlin/os/blazen/jessica/core/JessicaCore.kt
package os.blazen.jessica.core

/**
 * Idiomatic Kotlin façade over [JessicaCoreNative]. UI code should only
 * ever talk to this class; the JNI surface stays internal.
 *
 * M0 fallback: when [JessicaCoreNative.LIB_AVAILABLE] is false, the
 * façade runs a tiny pure-Kotlin intent matcher so the UI can still
 * exercise the API contract end-to-end on a dev laptop.
 *
 * M1: when LIB_AVAILABLE flips true, [matchIntent] needs a JSON parser
 * for the FFI's serde output — add kotlinx.serialization to `:core`'s
 * Gradle deps and wire it here. The contract under test (intent names,
 * params, confirm levels) doesn't change.
 */
class JessicaCore private constructor(
    private val handle: Long,
    private val native: Boolean,
) {

    private val fallback = if (!native) PureKotlinIntents() else null

    /** Load the YAML intent catalogue. Returns true on success. */
    fun loadIntents(yaml: String): Boolean {
        if (native) {
            val rc = JessicaCoreNative.nativeLoadIntents(handle, yaml.toByteArray(Charsets.UTF_8))
            return rc == JESSICA_OK
        }
        return fallback!!.load(yaml)
    }

    /** Match a transcript against the loaded catalogue. Returns null when no intent fires. */
    fun matchIntent(transcript: String, language: String): IntentMatch? {
        if (native) {
            JessicaCoreNative.nativeMatchIntent(handle, transcript, language)
                ?: return null
            throw NotImplementedError(
                "M1: parse the FFI JSON output here once kotlinx.serialization is on :core's classpath."
            )
        }
        return fallback!!.match(transcript, language)
    }

    /** How many intents are currently loaded. */
    fun intentCount(): Long {
        if (native) return JessicaCoreNative.nativeIntentCount(handle)
        return fallback!!.count().toLong()
    }

    /** Release the underlying handle. Idempotent. */
    fun close() {
        if (native) JessicaCoreNative.nativeFree(handle)
    }

    companion object {
        fun create(): JessicaCore {
            val useNative = JessicaCoreNative.LIB_AVAILABLE
            val handle = if (useNative) JessicaCoreNative.nativeNew() else 0L
            return JessicaCore(handle = handle, native = useNative)
        }
    }
}

internal const val JESSICA_OK = 0
internal const val JESSICA_ERR_BAD_HANDLE = -1
internal const val JESSICA_ERR_BAD_UTF8 = -2
internal const val JESSICA_ERR_BAD_INPUT = -3
internal const val JESSICA_ERR_PANIC = -99
